contracts/bridge/contracts/common/trust/LightClientBridge.sol

```diff
@@ -0,0 +1,662 @@
+     * @param threshold initial threshold of guard set
+     * @param validatorSetId initial validator set id
+     * @param validatorSetLen length of initial validator set
+     * @param validatorSetRoot initial validator set merkle tree root
+    */
+    constructor(
+        bytes32 network,
+        address payable slashVault,
+        address[] memory guards,
+        uint256 threshold,
+        uint256 validatorSetId,
+        uint256 validatorSetLen,
+        bytes32 validatorSetRoot
+    ) public GuardRegistry(network, guards, threshold) {
+        SLASH_VAULT = slashVault;
+        _updateValidatorSet(validatorSetId, validatorSetLen, validatorSetRoot);
+    }
+
+    /* Public Functions */
+
+    function getFinalizedChainMessagesRoot() external view returns (bytes32) {
+        return latestChainMessagesRoot;
+    }
+
+    function getFinalizedBlockNumber() external view returns (uint256) {
+        return latestBlockNumber;
+    }
+
+    function validatorBitfield(uint256 id) external view returns (uint256[] memory) {
+        return validationData[id].validatorClaimsBitfield;
+    }
+
+    function requiredNumberOfValidatorSigs() public view returns (uint256) {
+        if (validatorSetLen < 36) {
+            return validatorSetLen * 2 / 3 + 1;
+        }
+        return 25;
+    }
+
+    function createRandomBitfield(uint256 id)
+        public
+        view
+        returns (uint256[] memory)
+    {
+        ValidationData storage data = validationData[id];
+        return _createRandomBitfield(data);
+    }
+
+    function _createRandomBitfield(ValidationData storage data)
+        internal
+        view
+        returns (uint256[] memory)
+    {
+        require(data.blockNumber > 0, "Bridge: invalid id");
+        return
+            randomNBitsWithPriorCheck(
+                getSeed(data.blockNumber),
+                data.validatorClaimsBitfield,
+                requiredNumberOfValidatorSigs(),
+                validatorSetLen
+            );
+    }
+
+    function createInitialBitfield(uint256[] calldata bitsToSet, uint256 length)
+        external
+        pure
+        returns (uint256[] memory)
+    {
+        return createBitfield(bitsToSet, length);
+    }
+
+    function verify_messages_proof(
+        bytes32 outboundLaneDataHash,
+        bytes32 inboundLaneDataHash,
+        uint256 chain_pos,
+        uint256 lane_pos,
+        bytes calldata proof
+    ) external view returns (bool) {
+        bytes32 lane_hash = hash(LaneData(outboundLaneDataHash, inboundLaneDataHash));
+        return validate_messages_match_root(lane_hash, chain_pos, lane_pos, proof);
+    }
+
+    function verify_messages_delivery_proof(
+        bytes32 outboundLaneDataHash,
+        bytes32 inboundLaneDataHash,
+        uint256 chain_pos,
+        uint256 lane_pos,
+        bytes calldata proof
+    ) external view returns (bool) {
+        bytes32 lane_hash = hash(LaneData(outboundLaneDataHash, inboundLaneDataHash));
```

Comment:
OK, we used `inboundLaneDataHash` here, but do we really need `inboundLaneDataHash` here.

Reply:
same reason, https://github.com/darwinia-network/darwinia-bridge-sol/pull/77#discussion_r747188571

Reply:
As we discussed, outboundLaneDataHash/inboundLaneDataHash here is used for proving and should be part of proof.

Use OutboundLaneData and InboundLaneData as leaf of sparse merkle tree.